torchrl/collectors/distributed/distributed_collector.py

```diff
@@ -0,0 +1,232 @@
+
+    def stop(self):
+        """Stop all remote collectors"""
+        for w in self.remote_collectors():
+            w.__ray_terminate__.remote()
+
+    def __iter__(self) -> Iterator[TensorDictBase]:
+        if self.communication == "sync":
+            return self.sync_iterator()
+        else:
+            return self.async_iterator()
+
+    def sync_iterator(self) -> Iterator[TensorDictBase]:
+
+        while self.collected_frames < self.total_frames:
+
+            # Broadcast agent weights
+            self.local_collector().update_policy_weights_()
+            state_dict = {"policy_state_dict": self._local_collector.policy.state_dict()}
+            state_dict = ray.put(state_dict)
+            for e in self.remote_collectors():
+                e.load_state_dict.remote(**{"state_dict": state_dict, "strict": False})
+
+            # Ask for batches to all remote workers.
+            pending_samples = [e._iterator_step.remote() for e in self.remote_collectors()]
+
+            # Wait for all rollouts
+            samples_ready = []
+            while len(samples_ready) < self.num_collectors:
+                samples_ready, samples_not_ready = ray.wait(
+                    pending_samples, num_returns=len(pending_samples), timeout=0.001)
+
+            # Retrieve and concatenate Tensordicts
+            out_td = []
+            for r in pending_samples:
+                rollouts = ray.get(r)
+                ray.internal.free(r)
+                out_td.append(rollouts)
+            if len(rollouts.batch_size):
+                out_td = torch.stack(out_td)
+            else:
+                out_td = torch.cat(out_td)
+
+            self.collected_frames += out_td.numel()
+
+            yield out_td
+
+    def async_iterator(self) -> Iterator[TensorDictBase]:
+
+        pending_tasks = {}
+        for w in self.remote_collectors():
+            future = w._iterator_step.remote()
+            pending_tasks[future] = w
+
+        while self.collected_frames < self.total_frames:
+
+            if not len(list(pending_tasks.keys())) == len(self.remote_collectors()):
+                raise RuntimeError("Missing pending tasks, something went wrong")
+
+            # Wait for first worker to finish
+            wait_results = ray.wait(list(pending_tasks.keys()))
+            future = wait_results[0][0]
+            w = pending_tasks.pop(future)
+
+            # Retrieve single rollouts
+            out_td = ray.get(future)
+            ray.internal.free(future)
+            self.collected_frames += out_td.numel()
+
+            # Update agent weights
+            self.local_collector().update_policy_weights_()
+            state_dict = {"policy_state_dict": self._local_collector.policy.state_dict()}
+            state_dict = ray.put(state_dict)
+            w.load_state_dict.remote(**{"state_dict": state_dict, "strict": False})
+
+            # Schedule a new collection task
+            future = w._iterator_step.remote()
```

Comment:
when we reach the end of the loop, we'll still have workers waiting to deliver their data right? Isn't there a risk that they will stay idle forever?

Reply:
Yes, I will add some code to make sure remaining tasks are cancelled or waited out.